src/jltypes.c

```diff
@@ -2582,23 +1287,28 @@ void jl_reinstantiate_inner_types(jl_datatype_t *t) // can throw!
 
     for (j = 0; j < partial_inst.len; j++) {
         jl_datatype_t *ndt = (jl_datatype_t*)partial_inst.items[j];
-        assert(ndt->name->primary == (jl_value_t*)t);
+        assert(jl_unwrap_unionall(ndt->name->wrapper) == (jl_value_t*)t);
         for (i = 0; i < n; i++)
             env[i * 2 + 1] = jl_svecref(ndt->parameters, i);
 
-        ndt->super = (jl_datatype_t*)inst_type_w_((jl_value_t*)t->super, env, n, &top, 1);
+        ndt->super = (jl_datatype_t*)instantiate_with((jl_value_t*)t->super, env, n, NULL, &top);
         jl_gc_wb(ndt, ndt->super);
     }
 
     if (t->name->names != jl_emptysvec) {
         for (j = 0; j < partial_inst.len; j++) {
             jl_datatype_t *ndt = (jl_datatype_t*)partial_inst.items[j];
-            assert(ndt->name->primary == (jl_value_t*)t);
             for (i = 0; i < n; i++)
                 env[i * 2 + 1] = jl_svecref(ndt->parameters, i);
 
-            ndt->types = inst_all(t->types, env, n, &top, /*check*/1);
-            jl_gc_wb(ndt, ndt->types);
+            int k;
+            if (ndt->types == NULL) {
```

Comment:
shouldn't this be `assert(ndt->types == NULL && "shouldn't have attempted to construct types svec while inside a typedef")`